openflow13/nx_util.go

```diff
@@ -0,0 +1,278 @@
+	"OXM_OF_ARP_TPA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_TPA, 4),
+	"OXM_OF_ARP_SHA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_SHA, 6),
+	"OXM_OF_ARP_THA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_SHA, 6),
+	"OXM_OF_IPV6_SRC":       newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_SRC, 16),
+	"OXM_OF_IPV6_DST":       newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_DST, 16),
+	"OXM_OF_IPV6_FLABEL":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_FLABEL, 4),
+	"OXM_OF_ICMPV6_TYPE":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ICMPV6_TYPE, 1),
+	"OXM_OF_ICMPV6_CODE":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ICMPV6_CODE, 1),
+	"OXM_OF_IPV6_ND_TARGET": newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_ND_TARGET, 16),
+	"OXM_OF_IPV6_ND_SLL":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_ND_SLL, 6),
+	"OXM_OF_IPV6_ND_TLL":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_ND_TLL, 6),
+	"OXM_OF_MPLS_LABEL":     newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_MPLS_LABEL, 4),
+	"OXM_OF_MPLS_TC":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_MPLS_TC, 1),
+	"OXM_OF_MPLS_BOS":       newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_MPLS_BOS, 1),
+	"OXM_OF_PBB_ISID":       newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_PBB_ISID, 3),
+	"OXM_OF_TUNNEL_ID":      newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_TUNNEL_ID, 8),
+	"OXM_OF_IPV6_EXTHDR":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_EXTHDR, 2),
+}
+
+// oxxFieldHeaderWildMap is map to find target field header with mask using an OVS known OXX field name
+var oxxFieldHeaderWildMap map[string]*MatchField
+
+func init() {
+	oxxFieldHeaderWildMap = make(map[string]*MatchField)
+	for k, v := range oxxFieldHeaderMap {
+		field := &MatchField{
+			Class:   v.Class,
+			Field:   v.Field,
+			HasMask: true,
+			Length:  v.Length * 2,
+		}
+		oxxFieldHeaderWildMap[k] = field
+	}
+}
+
+func FindFieldHeaderByName(fieldName string, hasMask bool) (*MatchField, error) {
+	var err = fmt.Errorf("failed to find header by name %s", fieldName)
+	fieldKey := strings.ToUpper(fieldName)
+	var fieldsMap map[string]*MatchField
+	if hasMask {
+		fieldsMap = oxxFieldHeaderWildMap
+	} else {
+		fieldsMap = oxxFieldHeaderMap
+	}
+	field, found := fieldsMap[fieldKey]
+	if !found {
+		return nil, err
+	}
+
+	return field, nil
+}
+
+// Encode the range to a uint16 number
+func encodeOfsNbitsStartEnd(start uint16, end uint16) uint16 {
+	return (start << 6) + (end - start)
+}
+
+// Encode the range to a uint16 number
+// ofs is the start pos, nBits is the count of the range
+func encodeOfsNbits(ofs uint16, nBits uint16) uint16 {
+	return ofs<<6 | (nBits - 1)
+}
+
+func encodeOfs(ofsNbits uint16) uint16 {
```

Comment:
I think this is a decode function, not an encode one.

Reply:
Done.